### PR TITLE
[inductor] Generate indirect_indexing checks even if optimized out

### LINKED ISSUE #100831: [torch.compile] skips the check of `num_embeddings`
### 🐛 Describe the bug

`torch.compile` skips the check of `num_embeddings`, making the invalid model succeed

```py
import torch

torch.manual_seed(420)

x = torch.randint(0, 100, (1, 10))

class Model(torch.nn.Module):

    def __init__(self, input_size, num_embeddings):
        super(Model, self).__init__()
        self.embedding = torch.nn.Embedding(num_embeddings=num_embeddings, embedding_dim=input_size)

    def forward(self, input_ids):
        embeddings = self.embedding(input_ids)
        return embeddings

input_size = 1
num_embeddings = 1
func = Model(input_size, num_embeddings).to('cpu')

with torch.no_grad():
    func.train(False)

    jit_func = torch.compile(func)
    res2 = jit_func(x)
    print(res2)

    res1 = func(x) # without jit
    # IndexError: index out of range in self
```

### Versions

<details>
<summary>Click to expand</summary>

```
Collecting environment information...
PyTorch version: 2.1.0.dev20230503+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.1 LTS (x86_64)
GCC version: (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0
Clang version: 14.0.0-1ubuntu1
CMake version: Could not collect
Libc version: glibc-2.35

Python version: 3.9.16 (main, Mar  8 2023, 14:00:05)  [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.19.5-051905-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 11.5.119
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3060
Nvidia driver version: 510.108.03
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   46 bits physical, 48 bits virtual
Byte Order:                      Little Endian
CPU(s):                          24
On-line CPU(s) list:             0-23
Vendor ID:                       GenuineIntel
Model name:                      12th Gen Intel(R) Core(TM) i9-12900K
CPU family:                      6
Model:                           151
Thread(s) per core:              2
Core(s) per socket:              16
Socket(s):                       1
Stepping:                        2
CPU max MHz:                     6700.0000
CPU min MHz:                     800.0000
BogoMIPS:                        6374.40
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault cat_l2 invpcid_single cdp_l2 ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdt_a rdseed adx smap clflushopt clwb intel_pt sha_ni xsaveopt xsavec xgetbv1 xsaves split_lock_detect avx_vnni dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req hfi umip pku ospke waitpkg gfni vaes vpclmulqdq tme rdpid movdiri movdir64b fsrm md_clear serialize pconfig arch_lbr ibt flush_l1d arch_capabilities
Virtualization:                  VT-x
L1d cache:                       640 KiB (16 instances)
L1i cache:                       768 KiB (16 instances)
L2 cache:                        14 MiB (10 instances)
L3 cache:                        30 MiB (1 instance)
NUMA node(s):                    1
NUMA node0 CPU(s):               0-23
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Not affected
Vulnerability Retbleed:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected

Versions of relevant libraries:
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.24.2
[pip3] pytorch-triton==2.1.0+7d1a95b046
[pip3] torch==2.1.0.dev20230503+cu118
[pip3] torchaudio==2.1.0.dev20230503+cu118
[pip3] torchvision==0.16.0.dev20230503+cu118
[conda] numpy                     1.24.2                   pypi_0    pypi
[conda] pytorch-triton            2.1.0+7d1a95b046          pypi_0    pypi
[conda] torch                     2.1.0.dev20230503+cu118          pypi_0    pypi
[conda] torchaudio                2.1.0.dev20230503+cu118          pypi_0    pypi
[conda] torchvision               0.16.0.dev20230503+cu118          pypi_0    pypi
```

</details>

cc @ezyang @soumith @msaroufim @wconstab @ngimel @bdhirsh @anijain2305 @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### LINKED ISSUE #100878: Inductor guesses that an indirect indexing over a dimension of size 1 is correct
### 🐛 Describe the bug

The following incorrect program:
```python
import torch

@torch.compile
def f(x, y):
    return x[y]


x = torch.randn(1, device="cuda")
y = torch.ones(7, device="cuda")
```
generates the following code:

```python
def triton_(in_ptr0, in_ptr1, out_ptr0, xnumel, XBLOCK : tl.constexpr):                                                                                                                                              
    xnumel = 7                                                                                                                                                                                                       
    xoffset = tl.program_id(0) * XBLOCK                                                                                                                                                                              
    xindex = xoffset + tl.arange(0, XBLOCK)[:]                                                                                                                                                                       
    xmask = xindex < xnumel                                                                                                                                                                                          
    x0 = xindex                                                                                                                                                                                                      
    tmp0 = tl.load(in_ptr0 + (0))                                                                                                                                                                                    
    tmp1 = tl.broadcast_to(tmp0, [XBLOCK])                                                                                                                                                                           
    tmp2 = tl.load(in_ptr1 + (0))                                                                                                                                                                                    
    tmp3 = tl.broadcast_to(tmp2, [XBLOCK])                                                                                                                                                                           
    tl.store(out_ptr0 + (tl.broadcast_to(x0, [XBLOCK])), tmp3, xmask) 
```
TorchInductor assumes that, since you're indexing over a dimension with size 1, you want that element. This transformation is just sound for correct programs. Is that still alright?


### Versions

master

cc @ezyang @soumith @msaroufim @wconstab @ngimel @bdhirsh @anijain2305 @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #101602
* __->__ #100895
* #101077

Fixes #100831, fixes #100878

Previously `gen_assert_indirect_indexing` was only called on the index
expressions passed to `ops.load` and `ops.store` which means if the
variable is optimized out during lowering, we never generate the
assert. This instead makes `ops.indirect_indexing` eagerly generate
the assert statement, whether or not it will be used.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

File: torch/_inductor/codegen/triton.py [hunk 1]
```diff
@@ -1050,45 +1050,25 @@ def mask_loads(self, mask):
         finally:
             self._load_mask = prior
 
-    def gen_assert_indirect_indexing(self, buffer, original_index, mask):
-        if mask == "None":
-            return
-        body = self.current_node._body
-        indirect_size = dict(zip(body.indirect_vars, body.indirect_max_sizes))
```
Comment (lezcano) [inline torch/_inductor/codegen/triton.py:1057]: You can then also revert the code I added that populates ìndirect_max_sizes` and similar structures.

File: torch/_inductor/codegen/triton.py [hunk 2]
```diff
@@ -1050,45 +1050,25 @@ def mask_loads(self, mask):
...
+            mask_vars = list(var.mask_vars)
+            if mask_vars:
+                mask = (
+                    f"{mask_vars[0]}"
+                    if len(mask_vars) == 1
+                    else f"({' & '.join(mask_vars)})"
+                )
+                cond = f"({cond}) | ~{mask}"
+            line = f'tl.device_assert({cond}, "index out of bounds: {cond_print}")'
+            self.cse.generate(self.compute, line, assignment=False)
```
Comment (lezcano) [inline torch/_inductor/codegen/triton.py:1069]: Is it alright to always generate this in `self.compute`? Can't the load be in `self.load`, which happens before `self.compute`?

**Review by lezcano — COMMENTED** [review]
I think that an index variable is not often used in different expressions, so that should be fine.  We should benchmark this though.

I like how much more self-contained the implementation is.

**Comment by lezcano** [comment]
The issue with repeated asserts could potentially be solved via some mechanism similar to `DeferredLine`, but rather than removing the line, you could have a way to swap it.

File: torch/_inductor/codegen/triton.py (hunk 2 above)
Comment (peterbell10) [reply torch/_inductor/codegen/triton.py:1071]: Indirect loads are always in the `compute` buffer:
https://github.com/pytorch/pytorch/blob/660a0d8622d79fdc1a2b9a7102a5ecd33096c2ec/torch/_inductor/codegen/common.py#L717-L719

File: torch/_inductor/codegen/triton.py (hunk 1 above)
Comment (peterbell10) [reply torch/_inductor/codegen/triton.py:1057]: I spotted a problem with this. Theoretically you can have one maximum under one set of masks, and another under a different set of masks. Not an insurmountable problem, but I suspect it wouldn't be worth the complexity.

**Comment by peterbell10** [comment]
> The issue with repeated asserts could potentially be solved via some mechanism similar to `DeferredLine`, but rather than removing the line, you could have a way to swap it.

I spotted a problem with this. Theoretically you can have one maximum under one set of masks, and another under a different set of masks. Not an insurmountable problem, but I suspect it wouldn't be worth the complexity.

**Comment by jgong5** [comment]
Instead of assert multiple times, should we reduce the multiple indirect indexing with different sizes into a single assertion with the smallest size?

**Comment by lezcano** [comment]
>Instead of assert multiple times, should we reduce the multiple indirect indexing with different sizes into a single assertion with the smallest size?

This is what we were doing before. Also, I think that if we keep the `indirect_max_sizes` list, that should be populated before we generate this assertions, so we should still be able to do that with this approach.

File: torch/_inductor/codegen/triton.py [hunk 3]
```diff
@@ -1050,45 +1052,59 @@ def mask_loads(self, mask):
...
+            # An assertion line may have been written already, if so just
+            # update the max size.
+            size_map = self.indirect_max_sizes
+            map_key = (var_str, mask)
+            existing_size = size_map.get(map_key)
+            if existing_size is None:
+                line = 'tl.device_assert({cond}, "index out of bounds: {cond_print}")'
+                size_map[map_key] = size
+                self.compute.writeline(
+                    IndirectAssertLine(line, var_str, mask, size_map)
```
Comment (peterbell10) [inline torch/_inductor/codegen/triton.py:1102]: Okay, here is a completely different strategy for de-duplicating the assert lines.  I delay the final formatting of `cond` and `cond_print` until the entire IR has has been executed, and accumulate size information in `TritonKernel.indirect_max_sizes`. This mapping is keyed by variable name and by mask, so it correctly handles masked loads as well.

This also improves upon the current `main` branch which sometimes generates multiple asserts for the same variable. For example, this python code 
```python
import torch

def f(x, y, ind):
    return x[ind] + y[ind]

x = torch.randn(4, 3, device="cuda")
y = torch.randn(2, 3, device="cuda")
ind = torch.zeros(2, 2, dtype=torch.int64, device="cuda")
cfn = torch.compile(f, dynamic=True)
cfn(x, y, ind)
```

Generates two triton asserts on main,
```python
    tmp0 = tl.load(in_ptr0 + (x1), xmask)
    tl.device_assert((((0 <= tmp0) & (tmp0 < 4))) | ~xmask, "index out of bounds: 0 <= tmp0 < 4")
    tmp1 = tl.load(in_ptr1 + (x0 + (3*tmp0)), xmask)
    tl.device_assert((((0 <= tmp0) & (tmp0 < 2))) | ~xmask, "index out of bounds: 0 <= tmp0 < 2")
    tmp2 = tl.load(in_ptr2 + (x0 + (3*tmp0)), xmask)
```


But with this version it generates only one:
```python
    tl.device_assert(((0 <= tmp0) & (tmp0 < 2)) | ~xmask, "index out of bounds: 0 <= tmp0 < 2")
    tmp1 = tl.load(in_ptr1 + (x0 + (3*tmp0)), xmask)
    tmp2 = tl.load(in_ptr2 + (x0 + (3*tmp0)), xmask)
```
So it doesn't work when we generate the I don't really know how much value this is adding.

File: torch/_inductor/codegen/triton.py [hunk 4]
```diff
@@ -1050,45 +1056,70 @@ def mask_loads(self, mask):
...
+                    else f"({' & '.join(str(v) for v in mask_vars)})"
+                )
+
+            # tl.device_assert doesn't work for constexpr values, and we can't
+            # tell from here if a var is constexpr or not, so promote everything
+            var_str = str(
+                self.cse.generate(
+                    self.compute, f"triton_helpers.promote_to_tensor({var})"
+                )
+            )
```
Comment (lezcano) [inline torch/_inductor/codegen/triton.py:1101]: This is not going to do as per https://github.com/pytorch/pytorch/pull/100880#issuecomment-1540494933. Now, I think that most the "indirect indexing via scalars" (aka direct indexing) we do is not tagged as indirect_indexing, but I guess that there are some times where this is not the case and we do things like:
```python
tmp0 = 0
tl.load(in_ptr0 + tmp0, ...)
```
I don't know when is it that we generate this sort of code though.

File: torch/_inductor/codegen/common.py [hunk 5]
```diff
@@ -743,7 +742,8 @@ def inner(*args, **kwargs):
 
             @staticmethod
             def indirect_indexing(index_var, size):
-                return sympy_symbol(str(index_var))
+                # Skip CSE since this doesn't return an expression
+                return self.indirect_indexing(index_var, size)
```
Comment (vfdev-5) [inline torch/_inductor/codegen/common.py:746]: Peter, where comes `self` here ?

**Review by lezcano — APPROVED** [review]
Great stuff!

File: torch/_inductor/codegen/triton.py [hunk 6]
```diff
@@ -1050,45 +1056,70 @@ def mask_loads(self, mask):
...
-                size = print_min(sizes)
-            # The conditions need to be in parens because of Python's operator precedence.
-            # It'd be less # error-prone to use and/or/not, which is suported by triton
-            cond = f"((0 <= {var}) & ({var} < {size}))"
-            cond_print = f"0 <= {var} < {size}"
-            if not isinstance(original_index, sympy.Integer):
-                var_mask = f"({mask})" if "&" in mask else mask
-                var_mask = f" | ~{var_mask}"
+    def indirect_indexing(self, var, size):
+        class IndirectAssertLine(DeferredLineBase):
```
[the remark this answers is not in the document]
Comment (lezcano) [reply torch/_inductor/codegen/triton.py:1060]: nvm, I see you're already doing this.

[the remark this answers is not in the document]
Comment (peterbell10) [reply torch/_inductor/codegen/triton.py:1060]: That's already handled below:
```
        if config.triton.assert_indirect_indexing and torch.version.hip is None:
```

File: torch/_inductor/triton_helpers.py [hunk 7]
```diff
@@ -4,11 +4,15 @@
 TRITON_HAS_REDUCE = hasattr(tl, "reduce")
 
 
+@triton.jit
+def promote_to_tensor(x):
+    # Addition promotes to tensor for us
+    return x + tl.zeros((1,), tl.int1)
```
[the remark this answers is not in the document]
Comment (peterbell10) [reply torch/_inductor/triton_helpers.py:10]: Doesn't seem like either of these work
```
>>> tl.zeros((), tl.int1)
TypeError('0d block_type is forbidden')

>>> tl.broadcast(1.0, tl.zeros((1,), tl.int1)
AttributeError("'constexpr' object has no attribute 'type'")
```

**Comment by ngimel** [comment]
Can we delay merging this until the next PR is ready to merge (is it?) so that we don't have to live with the perf regression for long?

**Review by ngimel — APPROVED** [review]

**Pull request closed without merging**